### ISSUE TITLE
Higher serial speed don't work in :DIR mode

### ISSUE THREAD

**Issue by hubertushirsch** [Kris-Sekula/EPROM-EMU-NG#12]
High Kris,

I had previously done tests with higher serial data rates and discussed them on groups.io.
I had successfully tested the FW/SW (rc3/rc3) with 115200 (default), 250000, 500000 and 1000000bit/s.
Higher speeds lead to shorter transmission times.
This no longer works in the current rc11 firmware/rc9 software.
When transferring with the old method (spi=yes => :SBN) it works as before.
With SRAM only mode (spi=no => :DIR), the transmission time is shortened when increasing from 115200 to 250000bit/s as expected.
At 500000 and 1000000bit/s it now takes 64s.

I haven't investigated the problem yet.
I suspect it is due to the serial receiving algorithm in the firmware (buffer problem?).
At first glance, the send routine in the Python script is identical except for the start command (:SBN vs. :DIR)

Thus, even with "SRAM only" (spi=no), a higher transmission rate would have a better effect than different algorithms for spi=no and spi=yes.

Ralf

**Comment by Kris-Sekula** [comment]
Hey Ralf

Yeah, that is intentional, unfortunately. I was having issues with a few of the Nano boards with CH340 running at 1000000bit/s, so by default, I didn't push the code to run at that speed. But I still wanted to make the transfer faster so I optimize the code based on the 115200 speed by writing directly to SRAM instead of buffering the data on the Nano. This gave me a boost, I think I "shaved" 1-2 sec in comparison to going via the buffer. This works well but only at 115200, if you go faster the Nano doesn't have time to complete the write to SRAM before another byte arrives from the computer. Attached is a modified Python code that brings back the :SBN for all transfers, and will run at  1000000bit/s, don't forget to adjust the speed in the Arduino code as well.
[EPROM_EMU_NG_2.0rc9b.py.txt](https://github.com/Kris-Sekula/EPROM-EMU-NG/files/7574060/EPROM_EMU_NG_2.0rc9b.py.txt)

**Comment by hubertushirsch** [comment]
Hi Kris,

thanks for the answer.

115200 bit/s is not a law. It's just a limitation in the Arduino IDE's serial monitor.
Nobody restricts the transfer of data between two systems as quickly as the technical possibilities of the systems allow.

I have made measurements in all combinations, always with FW rc11 and SW rc9(b).
A 27512 EPROM with 80% real code and 20% 0xff is transmitted in each case.
Based on these measurements, everyone can decide which mode they prefer.

115200 :SBN  7,47s 
115200 :DIR  6,05s (saved 1,42s compared to 115200 :SBN)

250000 :SBN  4,27s (saved 3,2s to 115200 :SBN or 1,78s to 115200 :DIR)
250000 :DIR  2,94s

500000 :SBN  2,92s (saved 4,55s to 115200 :SBN or 3,13s to 115200 :DIR)
500000 :DIR  >60s

1000000 :SBN  2,31s (saved 5,16s to 115200 :SBN or 3,74s to 115200 :DIR)
1000000 :DIR  >60s

You see, with the same transfer rate :DIR is faster than :SBN, but :DIR does not work at higher transfer rates.
But :SBN at 250000 bit/s is already significantly faster than :DIR at 115200 bit/s.
If your goal is to save time, :SBN with a higher transmission rate is the more efficient solution.

Ralf

**Comment by Kris-Sekula** [comment]
Hey Ralf,

I don't dispute that faster transfer is better :) 100% with you on that.
The limitation is not Arduino IDE / or Arduino serial monitor. The only
reason I was sticking to 115200 was that a few of the Arduino Nano clone
modules that I used that came from China had a ch340 chip onboard and
although the chip is supposed to support higher speeds, out of 40 modules
that I used, 2-3 of them were failing to communicate at speeds over 115200,
hence I decided not to go over that speed.

Even at 115200 I was trying to find a way to speed up the transfer, and
assuming I only want to run at 115200, one way of speeding things up was to
stop buffering the data once it arrives in the emulator and send it
directly to the SRAM.

I encourage anyone who wants to get faster transfer time to use faster
speeds, but by default I ship the software and firmware configured at
115200 so that I know for sure it will work with any of my Nano modules out
there.

This is the beauty of "open source", you got the power to make a change
based on your needs :)

So as a conclusion, I feel we are both "correct", but thanks for pointing
out and reminding everyone that running the emulator at higher speeds is
possible and will give shorter transfer times.

Kris
